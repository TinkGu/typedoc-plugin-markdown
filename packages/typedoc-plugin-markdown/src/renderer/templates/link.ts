import * as path from 'path';

import settings from '../settings';
import { link } from '../tools/elements';
import { escapeChars } from '../tools/utils';

export function linkTemplate(label: string, to = '', escapeLabel = true) {
  label = escapeLabel ? escapeChars(label) : label;
  return link(label, getUrl(to));
}

function getUrl(to: string) {
  const { options, activeUrl } = settings;

  const urlPrefix = /^(http|ftp)s?:\/\//;
  if (urlPrefix.test(to)) {
    return to;
  } else {
    return options.publicPath
      ? getPublicUrl(options.publicPath, to)
      : getRelativeUrl(activeUrl, to);
  }
}

function getPublicUrl(publicPath: string, url: string) {
  return publicPath + url;
}

function getRelativeUrl(from: string, to: string) {
  const relative = path.relative(path.dirname(from), path.dirname(to));
  return path.join(relative, path.basename(to)).replace(/\\/g, '/');
}
