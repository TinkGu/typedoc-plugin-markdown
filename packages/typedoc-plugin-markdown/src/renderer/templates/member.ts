import { DeclarationReflection, ReferenceReflection } from 'typedoc';
import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { heading } from '../tools/elements';
import { escapeChars } from '../tools/utils';
import { declarationTemplate } from './declaration';
import { signatureTemplate } from './signature';

export function memberTemplate(
  model: DeclarationReflection | ReferenceReflection,
) {
  const options = settings.options;
  const md = new MarkdownBuilder();

  md.add(
    heading(
      3,
      options.namedAnchors
        ? `<a id="${model.anchor}" name="${model.anchor}"></a> `
        : '' + escapeChars(model.name),
    ),
  );

  if (model.signatures) {
    model.signatures.forEach((signature) => {
      md.add(signatureTemplate(signature));
    });
  } else {
    if (model.hasGetterOrSetter()) {
      if (model.getSignature) {
        md.add(signatureTemplate(model.getSignature, false, 'get'));
      }
      if (model.setSignature) {
        md.add(signatureTemplate(model.setSignature, false, 'set'));
      }
    } else {
      if (model instanceof ReferenceReflection && model.isReference) {
        const deep = model.tryGetTargetReflectionDeep();
        if (deep) {
          md.add(`Re-exports ${deep.name}`);
        }
      }

      if (model instanceof DeclarationReflection) {
        md.add(declarationTemplate(model));
      }
    }
  }

  return md.print();
}
