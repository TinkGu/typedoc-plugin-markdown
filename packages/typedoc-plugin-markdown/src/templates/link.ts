import * as path from 'path';
import { URL_PREFIX } from '../constants';
import { useSettings } from '../context';
import { link } from '../tools/elements';
import { escapeChars } from '../tools/utils';

export function linkTemplate(label: string, to = '', escapeLabel = true) {
  label = escapeLabel ? escapeChars(label) : label;
  return link(label, getUrl(to));
}

function getUrl(to: string) {
  const { activeUrl, publicPath } = useSettings();

  if (URL_PREFIX.test(to)) {
    return to;
  } else {
    return publicPath
      ? getPublicUrl(publicPath, to)
      : getRelativeUrl(activeUrl, to);
  }
}

function getPublicUrl(publicPath: string, url: string) {
  return publicPath + url;
}

function getRelativeUrl(from: string, to: string) {
  const relative = path.relative(path.dirname(from), path.dirname(to));
  return path.join(relative, path.basename(to)).replace(/\\/g, '/');
}
