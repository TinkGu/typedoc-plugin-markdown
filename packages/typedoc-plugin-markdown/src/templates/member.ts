import { DeclarationReflection, ReferenceReflection } from 'typedoc';
import { getContext } from '../context';
import { MarkdownBuilder } from '../tools/builder';
import { heading } from '../tools/elements';
import { escapeChars } from '../tools/utils';
import { declarationTemplate } from './declaration';
import { signatureTemplate } from './signature';

export function memberTemplate(
  model: DeclarationReflection | ReferenceReflection,
) {
  const { namedAnchors } = getContext();

  const md = new MarkdownBuilder();

  md.add(
    heading(
      3,
      namedAnchors
        ? `<a id="${model.anchor}" name="${model.anchor}"></a> `
        : '' + escapeChars(model.name),
    ),
  );

  if (model.signatures) {
    model.signatures.forEach((signature) => {
      md.add(signatureTemplate(signature));
    });
  } else {
    if (model.hasGetterOrSetter()) {
      if (model.getSignature) {
        md.add(signatureTemplate(model.getSignature, false, 'get'));
      }
      if (model.setSignature) {
        md.add(signatureTemplate(model.setSignature, false, 'set'));
      }
    } else {
      if (model instanceof ReferenceReflection && model.isReference) {
        const deep = model.tryGetTargetReflectionDeep();
        if (deep) {
          md.add(`Re-exports ${deep.name}`);
        }
      }

      if (model instanceof DeclarationReflection) {
        md.add(declarationTemplate(model));
      }
    }
  }

  return md.print();
}
