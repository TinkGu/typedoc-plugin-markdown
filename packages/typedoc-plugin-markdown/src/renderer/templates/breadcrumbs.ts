import { Reflection } from 'typedoc';
import { PageEvent } from 'typedoc/dist/lib/output/events';
import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { escapeChars } from '../tools/utils';
import { linkTemplate } from './link';

export function breadcrumbsTemplate(page: PageEvent) {
  const { activeUrl, globalsName, globalsFile, options } = settings;

  const md = new MarkdownBuilder();

  md.add(
    page.url === options.entryDocument
      ? page.project.name
      : linkTemplate(page.project.name, options.entryDocument),
  );

  if (!options.readme.endsWith('none')) {
    if (activeUrl === globalsFile) {
      md.add(globalsName);
    } else {
      md.add(linkTemplate(globalsName, globalsFile));
    }
  }

  const addBreadcrumb = (model: Reflection, md: MarkdownBuilder) => {
    if (model && model.parent) {
      addBreadcrumb(model.parent, md);

      if (activeUrl === model.url) {
        md.add(escapeChars(model.name));
      } else {
        md.add(linkTemplate(model.name, model.url));
      }
    }
  };

  addBreadcrumb(page.model, md);

  return md.print(' / ');
}
