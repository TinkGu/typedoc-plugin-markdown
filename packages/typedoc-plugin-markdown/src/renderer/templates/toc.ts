import {
  DeclarationReflection,
  ProjectReflection,
  ReflectionCategory,
  ReflectionGroup,
} from 'typedoc/dist/lib/models';

import settings from '../settings';
import { MarkdownBuilder } from '../tools/builder';
import { heading, unorderedList } from '../tools/elements';
import { linkTemplate } from './link';

export function tocTemplate(model: ProjectReflection | DeclarationReflection) {
  const options = settings.options;
  const md = new MarkdownBuilder();
  const isVisible = model.groups?.some((group) =>
    group.allChildrenHaveOwnDocument(),
  );
  if ((!this.hideInPageTOC && model.groups) || (isVisible && model.groups)) {
    if (!options.hideInPageTOC) {
      md.add(heading(2, 'Table of contents'));
    }
    model.groups
      ?.filter(
        (group) => !options.hideInPageTOC || group.allChildrenHaveOwnDocument(),
      )
      .forEach((group) => {
        if (group.categories) {
          group.categories.forEach((category) => {
            md.add(getGroup(category, category.title + ' ' + group.title));
          });
        } else {
          md.add(getGroup(group, group.title));
        }
      });
  }

  return md.print();
}

function getGroup(group: ReflectionGroup | ReflectionCategory, title: string) {
  const options = settings.options;
  const md = new MarkdownBuilder();
  md.add(heading(options.hideInPageTOC ? 2 : 3, title));
  const groupItems = group.children.map((child) =>
    linkTemplate(child.name, child.url),
  );
  md.add(unorderedList(groupItems));
  return md.print();
}
