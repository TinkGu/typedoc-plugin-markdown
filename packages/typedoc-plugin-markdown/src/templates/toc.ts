import {
  DeclarationReflection,
  ProjectReflection,
  ReflectionCategory,
  ReflectionGroup,
} from 'typedoc/dist/lib/models';
import { getContext } from '../context';
import { MarkdownBuilder } from '../tools/builder';
import { heading, unorderedList } from '../tools/elements';
import { linkTemplate } from './link';

export function tocTemplate(model: ProjectReflection | DeclarationReflection) {
  const { hideInPageTOC } = getContext();

  const md = new MarkdownBuilder();
  const isVisible = model.groups?.some((group) =>
    group.allChildrenHaveOwnDocument(),
  );
  if ((!this.hideInPageTOC && model.groups) || (isVisible && model.groups)) {
    if (!hideInPageTOC) {
      md.add(heading(2, 'Table of contents'));
    }
    model.groups
      ?.filter((group) => !hideInPageTOC || group.allChildrenHaveOwnDocument())
      .forEach((group) => {
        if (group.categories) {
          group.categories.forEach((category) => {
            md.add(getGroup(category, category.title + ' ' + group.title));
          });
        } else {
          md.add(getGroup(group, group.title));
        }
      });
  }

  return md.print();
}

function getGroup(group: ReflectionGroup | ReflectionCategory, title: string) {
  const { hideInPageTOC } = getContext();
  const md = new MarkdownBuilder();
  md.add(heading(hideInPageTOC ? 2 : 3, title));
  const groupItems = group.children.map((child) =>
    linkTemplate(child.name, child.url),
  );
  md.add(unorderedList(groupItems));
  return md.print();
}
