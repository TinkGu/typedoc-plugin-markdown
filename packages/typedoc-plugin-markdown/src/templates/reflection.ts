import { DeclarationReflection, PageEvent, ProjectReflection } from 'typedoc';

import { typeTemplate } from './type';
import { MarkdownBuilder } from '../tools/builder';
import { heading, unorderedList } from '../tools/elements';
import { commentsTemplate } from './comments';
import { hierarchyTemplate } from './hierarchy';

import { reflectionPathTemplate } from './reflection-path';
import { typeParameterTableTemplate } from './table.type-parameter';
import { signatureTemplate } from './signature';
import { tocTemplate } from './toc';
import { groupsTemplate } from './groups';
import { breadcrumbsTemplate } from './breadcrumbs';
import { useSettings } from '../context';

export function reflectionTemplate(page: PageEvent<DeclarationReflection>) {
  const { hideBreadcrumbs } = useSettings();

  const md = new MarkdownBuilder();

  if (!hideBreadcrumbs) {
    md.add(breadcrumbsTemplate(page));
  }

  md.add(reflectionPathTemplate(page.model));

  if (page.model.comment) {
    md.add(commentsTemplate(page.model.comment));
  }

  if (page.model.typeParameters) {
    md.add(heading(2, 'Type parameters'));
    md.add(typeParameterTableTemplate(page.model.typeParameters));
  }

  if (page.model.typeHierarchy && page.model.typeHierarchy.next) {
    md.add(heading(2, 'Hierarchy'));
    md.add(hierarchyTemplate(page.model.typeHierarchy));
  }

  if (page.model.implementedTypes) {
    md.add(heading(2, 'Implements'));
    md.add(
      unorderedList(
        page.model.implementedTypes.map((implementedType) =>
          typeTemplate(implementedType),
        ),
      ),
    );
  }

  if (page.model.signatures) {
    md.add(heading(2, 'Callable'));
    page.model.signatures.forEach((signature) => {
      md.add(heading(3, signature.name));
      md.add(signatureTemplate(signature));
    });
  }

  md.add(tocTemplate(page.model));

  if (page.model.groups) {
    md.add(groupsTemplate(page.model.groups));
  }

  return md.print();
}
