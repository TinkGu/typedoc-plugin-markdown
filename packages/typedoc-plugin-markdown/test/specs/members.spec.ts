import { DeclarationReflection } from 'typedoc';
import * as linkTemplate from '../../src/templates/link';
import * as memberTemplate from '../../src/templates/member';
import { reflectionTemplate } from '../../src/templates/reflection';
import * as sourcesTemplate from '../../src/templates/sources';
import * as tocTemplate from '../../src/templates/toc';
import { formatContents } from '../../src/tools/utils';
import { TestApp } from '../test-app';

describe(`Members:`, () => {
  let testApp: TestApp;

  beforeAll(() => {
    testApp = new TestApp(['members.ts']);
    testApp.bootstrap();
    jest.spyOn(tocTemplate, 'tocTemplate').mockReturnValue('[TOC]');
    jest.spyOn(linkTemplate, 'linkTemplate').mockReturnValue('[LINK]');
    jest.spyOn(sourcesTemplate, 'sourcesTemplate').mockReturnValue('[SOURCES]');
  });

  afterAll(() => {
    jest.restoreAllMocks();
    testApp.cleanup();
  });

  describe(`(members)`, () => {
    let memberSpy: jest.SpyInstance;
    beforeAll(() => {
      memberSpy = jest
        .spyOn(memberTemplate, 'memberTemplate')
        .mockReturnValue('[MEMBER]');
    });
    afterAll(() => {
      memberSpy.mockRestore();
    });
    test(`should compile class members'`, () => {
      expect(
        formatContents(
          reflectionTemplate(
            testApp.findReflection('ClassWithAccessorMembers'),
          ),
        ),
      ).toMatchSnapshot();
    });
  });

  describe(`(member)`, () => {
    test(`should compile declaration members'`, () => {
      expect(
        formatContents(
          memberTemplate.memberTemplate(
            testApp.findReflection('declarationMember'),
          ),
        ),
      ).toMatchSnapshot();
    });

    test(`should compile a signature members'`, () => {
      expect(
        formatContents(
          memberTemplate.memberTemplate(
            testApp.findReflection('signatureMember'),
          ),
        ),
      ).toMatchSnapshot();
    });

    test(`should compile members with getter'`, () => {
      expect(
        formatContents(
          memberTemplate.memberTemplate(
            testApp
              .findReflection('ClassWithAccessorMembers')
              .findReflectionByName('getter') as DeclarationReflection,
          ),
        ),
      ).toMatchSnapshot();
    });

    test(`should compile members with setter'`, () => {
      expect(
        formatContents(
          memberTemplate.memberTemplate(
            testApp
              .findReflection('ClassWithAccessorMembers')
              .findReflectionByName('setter') as DeclarationReflection,
          ),
        ),
      ).toMatchSnapshot();
    });
  });
});
