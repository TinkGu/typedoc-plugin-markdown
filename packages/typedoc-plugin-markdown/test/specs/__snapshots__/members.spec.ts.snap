// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Members: (member) should compile a signature members' 1`] = `
"### signatureMember

▸ **signatureMember**(): \`void\`

#### Returns

\`void\`

[SOURCES]
"
`;

exports[`Members: (member) should compile declaration members' 1`] = `
"### declarationMember

• \`Let\` **declarationMember**: \`string\`

[SOURCES]
"
`;

exports[`Members: (member) should compile members with getter' 1`] = `
"### getter

• \`get\` **getter**(): \`string\`

#### Returns

\`string\`

[SOURCES]
"
`;

exports[`Members: (member) should compile members with setter' 1`] = `
"### setter

• \`set\` **setter**(\`value\`): \`void\`

#### Parameters

| Name | Type |
| :------ | :------ |
| \`value\` | \`string\` |

#### Returns

\`void\`

[SOURCES]
"
`;

exports[`Members: (members) should compile class members' 1`] = `
"[TOC]

## Constructors

[MEMBER]

## Properties

[MEMBER]

## Accessors

[MEMBER]

---

[MEMBER]
"
`;
